src/audio/openal/castleopenalsoundbackend.pas

```diff
@@ -50,10 +52,35 @@   TOpenALSoundBufferBackend = class(TSoundBufferBackendFromSoundFile)
     procedure ContextClose; override;
   end;
 
+  TOpenALStreamBuffersBackend = class (TSoundBufferBackendFromStreamedFile)
+  private
+    ALBuffers: array[0..3] of TALuint;
+    StreamedSFile: TStreamedSoundFile;
+    FeedThread: TOpenALStreamFeedThread;
+
+    function FillBuffer(ALBuffer: TALuint): Integer;
+  public
+    procedure FeedBuffers(ALSource: TALuint);
+    procedure StopFeedingBuffers;
+
+    procedure ContextOpenFromStreamedFile(const StreamedSoundFile: TStreamedSoundFile); override;
+    procedure ContextClose; override;
+  end;
+
+  TOpenALStreamFeedThread = class(TThread)
```

Comment:
Note that some platforms do not have working threads now. E.g. Nintendo Switch in the current implementation. Other weird platforms may have this limitation too :)

So the threading code should be surrounded by a clause like `{$ifdef CASTLE_SUPPORTS_THREADING}`. When it is not defined, the code should still work, just fallback to non-threading version, which means:

- Maybe read parts of files in the main thread. That is, do what you're currently doing, but not inside TThread.Execute. Only if it makes sense at all for some buffer sizes (we would need to read really small chunks of files to not cause a slowdown, I guess? And any file access is actually a bad idea in the middle of rendering... So this may be a stupid idea from my side.)

- Or just fallback to a version of playing without streaming.

The point is, game sound should work, even when the platform doesn't support streaming.

Reply:
> Note that some platforms do not have working threads now. E.g. Nintendo Switch in the current implementation. Other weird platforms may have this limitation too :)

No threads, I did not expect that :) Is this a limitation of FeePascal or Nintendo Switch?

Reply:
It's a limitation of my port to Nintendo Switch (the Nintendo Switch parts are in closed-source CGE version now, as Nintendo doesn't allow to publish anything about their API to general public).

It is 100% possible to make Pascal `TThread` on Nintendo Switch working, but it would be an additional work that I simply didn't do yet :)

Reply:
I think that introducing threads and streams of music could greatly improve performance on the Switch (in your games) just like on Android. After switching to music streaming, my test devices  are also much cooler.

Reply:
Right now on Switch we use (at Cat-astrophe Games) FMOD backend, https://github.com/castle-engine/castle-engine/wiki/FMOD . The OpenAL is not available on Nintendo Switch. While there is a "native Nintendo Switch backend" implemented in CGE (part of closed-source CGE for Nintendo Switch code), but FMOD is just more functional.

FMOD provides streaming in a simpler way (I did not implement it yet, but I remember seeing it in docs) -- you just pass a flag like "streaming" to some API call, and it handles the actual file loading + streaming underneath. So it deals with eventual threads by itself.

Reply:
P.S. This actually makes my earlier comment ("streaming must work without threads") less useful. I said that threads are not implemented on Nintendo Switch, and now I said that on Nintendo Switch we don't use OpenAL :)

So, this has low priority.  For safety I would prefer to have a path for platforms that don't support threads but support OpenAL, but it's not really important now in any practical place.

Reply:
Done in https://github.com/castle-engine/castle-engine/pull/127/commits/80c2989c7853a83cac9d36f54ab8a5588f666f8e